engine/src/main/java/org/hibernate/validator/internal/metadata/aggregated/PropertyCascadable.java

```diff
@@ -6,37 +6,38 @@
  */
 package org.hibernate.validator.internal.metadata.aggregated;
 
-import java.lang.annotation.ElementType;
-import java.lang.reflect.Method;
 import java.lang.reflect.Type;
 
 import org.hibernate.validator.internal.engine.path.PathImpl;
 import org.hibernate.validator.internal.engine.valueextraction.ValueExtractorManager;
 import org.hibernate.validator.internal.metadata.facets.Cascadable;
-import org.hibernate.validator.internal.util.ReflectionHelper;
+import org.hibernate.validator.internal.metadata.location.ConstraintLocation.ConstraintLocationKind;
+import org.hibernate.validator.internal.properties.Property;
+import org.hibernate.validator.internal.properties.javabean.JavaBeanField;
 
 /**
- * A {@link Cascadable} backed by a property getter of a Java bean.
+ * A {@link Cascadable} backed by a field of a Java bean.
  *
  * @author Gunnar Morling
+ * @author Marko Bekhta
  */
-public class GetterCascadable implements Cascadable {
+public class PropertyCascadable implements Cascadable {
 
-	private final Method method;
-	private final String propertyName;
+	private final Property property;
 	private final Type cascadableType;
 	private final CascadingMetaData cascadingMetaData;
+	private final ConstraintLocationKind constraintLocationKind;
 
-	GetterCascadable(Method method, CascadingMetaData cascadingMetaData) {
-		this.method = method;
-		this.propertyName = ReflectionHelper.getPropertyName( method );
-		this.cascadableType = ReflectionHelper.typeOf( method );
+	PropertyCascadable(Property property, CascadingMetaData cascadingMetaData) {
+		this.property = property;
+		this.cascadableType = property.getType();
 		this.cascadingMetaData = cascadingMetaData;
+		this.constraintLocationKind = property instanceof JavaBeanField ? ConstraintLocationKind.PROPERTY : ConstraintLocationKind.METHOD;
```

Comment:
We shouldn't have to rely on `JavaBeanField` here.

That will probably change if we reintroduce the difference between getters and properties. Frankly, I would really like to be able to follow the path of a getter and understand how it's specific.

If we end up still having to do that, I think we should have a `ConstrainedElementKind getConstrainedElementKind()` in `Constrainable` (instead of `Callable`). That might help for other spots too.